lib/awful/popup.lua

```diff
@@ -0,0 +1,434 @@
+--
+-- As demonstrated by this second example, it is also possible to use a widget
+-- as a parent object:
+--
+-- @DOC_awful_popup_position2_EXAMPLE@
+--
+-- @property preferred_positions
+-- @tparam table|string preferred_positions A position name or an ordered
+--  table of positions
+-- @see awful.placement.next_to
+-- @see awful.popup.preferred_anchors
+
+function popup:set_preferred_positions(pref_pos)
+    self._private.preferred_directions = pref_pos
+    set_position(self)
+end
+
+--- Set the preferred popup anchors relative to the parent.
+--
+-- The possible values are:
+--
+-- * front
+-- * middle
+-- * back
+--
+-- For details information, see the `awful.placement.next_to` documentation.
+--
+-- In this example, it is possible to see the effect of having a fallback
+-- preferred anchors when the popup would otherwise not fit:
+--
+-- @DOC_awful_popup_anchors_EXAMPLE@
+--
+-- @property preferred_anchors
+-- @tparam table|string preferred_anchors Either a single anchor name or a table
+--  ordered by priority.
+-- @see awful.placement.next_to
+-- @see awful.popup.preferred_positions
+
+function popup:set_preferred_anchors(pref_anchors)
+    self._private.preferred_anchors = pref_anchors
+    set_position(self)
+end
+
+--- The current position relative to the parent object.
+--
+-- If there is a parent object (widget, wibox, wibar, client or the mouse), then
+-- this property returns the current position. This is determined using
+-- `preferred_positions`. It is usually the preferred position, but when there
+-- isn't enough space, it can also be one of the fallback.
+--
+-- @property current_position
+-- @param string Either "left", "right", "top" or "bottom"
+
+function popup:get_current_position()
+    return self._private.current_position
+end
+
+--- Get the current anchor relative to the parent object.
+--
+-- If there is a parent object (widget, wibox, wibar, client or the mouse), then
+-- this property returns the current anchor. The anchor is the "side" of the
+-- parent object on which the popup is based on. It will "grow" in the
+-- opposite direction from the anchor.
+--
+-- @property current_anchor
+-- @tparam string current_anchor Either "front", "middle", "back"
+
+function popup:get_current_anchor()
+    return self._private.current_anchor
+end
+
+--- Move the wibox to a position relative to `geo`.
+-- This will try to avoid overlapping the source wibox and auto-detect the right
+-- direction to avoid going off-screen.
+--
+-- @param[opt=mouse] obj An object such as a wibox, client or a table entry
+--  returned by `wibox:find_widgets()`.
+-- @see awful.placement.next_to
+-- @see awful.popup.preferred_positions
+-- @see awful.popup.preferred_anchors
+-- @treturn table The new geometry
+function popup:move_next_to(obj)
+    if self._private.is_relative == false then return end
+
+    self._private.widget_geo = obj
+
+    if not obj then return end
+
+    if obj._apply_size_now then
+        obj:_apply_size_now(false)
```

Comment:
It is necessary because popups can be placed in relation to each other (like the notifications) and at some point it is necessary to collapse all the async stuff. But the async stuff are also necessary for some other corner case like the screenshot above. Without the async part, it can actually crash Awesome.